### PR TITLE
Enable Spark 1.4.0 local mode for Python 3 and Scala

### PR DESCRIPTION
**Pull request by parente**
* Build Scala kernel against Spark 1.4.0
* Install Scala kernel spec
* Configure Python3 paths for pyspark 1.4.0

Address issue #41 (and more).

### CODE REVIEW

**Comment by Carreau** [comment]
Does that make a Ptyhon3-spark kernel ? Because there is no-need for a Pyspark kernel, you can just import [findspark](https://github.com/minrk/findspark).

File: Dockerfile [hunk 1]
```diff
@@ -70,6 +99,10 @@ RUN cabal update && \
     ihaskell install && \
     rm -fr $(echo ~/.cabal/bin/* | grep -iv ihaskell) ~/.cabal/packages ~/.cabal/share/doc ~/.cabal/setup-exe-cache ~/.cabal/logs
 
+# Spark kernel spec
+RUN mkdir -p $HOME/.ipython/kernels/spark
```
Comment (rgbkrk) [inline Dockerfile]: Would it be appropriate to call this a Scala kernel instead or do you see this as dissimilar?

Comment (parente) [inline Dockerfile]: Good catch. That was a naming mixup on my part. I've forced pushed and corrected the directory to `scala` along with the comment.

**Comment by parente** [comment]
The Scala kernel with Spark bindings is not shipped in binary form so there's somewhat ugly buildstep in the Dockerfile for that.

The only config done to enable Python3 to use Spark in the container is setting the PYTHONPATH and SPARK_HOME. Since this is in a Docker image with a well-known filesystem structure, I didn't think the findspark machinery for multiple OSes and install locations was necessary.

File: Dockerfile (hunk 1 above)
Comment (parente) [inline Dockerfile]: Also corrected the kernel display name to language-first (Scala).

File: Dockerfile [hunk 2]
```diff
@@ -39,7 +66,9 @@ USER jovyan
 ENV HOME /home/jovyan
 ENV SHELL /bin/bash
 ENV USER jovyan
+ENV SPARK_HOME /usr/local/spark
 ENV PATH $CONDA_DIR/bin:$CONDA_DIR/envs/python2/bin:$PATH
+ENV PYTHONPATH $SPARK_HOME/python:$SPARK_HOME/python/lib/py4j-0.8.2.1-src.zip
```
Comment (rgbkrk) [inline Dockerfile]: Would it matter to append `:$PYTHONPATH` to the end here?

File: Dockerfile [hunk 3]
```diff
@@ -70,6 +99,10 @@ RUN cabal update && \
     ihaskell install && \
     rm -fr $(echo ~/.cabal/bin/* | grep -iv ihaskell) ~/.cabal/packages ~/.cabal/share/doc ~/.cabal/setup-exe-cache ~/.cabal/logs
 
+# Scala Spark kernel spec
+RUN mkdir -p $HOME/.ipython/kernels/scala
+COPY resources/kernel.json $HOME/.ipython/kernels/scala/
+
```
Comment (rgbkrk) [inline Dockerfile]: Looking good! I'll probably start tinkering with Scala when available here. :wink:

File: Dockerfile (hunk 2 above)
Comment (rgbkrk) [inline Dockerfile]: Ah, right, `PYTHONPATH` is for adding additional directories. Nevermind me.

**Comment by rgbkrk** [comment]
Thanks @parente!

**Comment by ellisonbg** [comment]
I am very much in favor of this. What about using latest Spark stable (1.4.1) though?

**Comment by parente** [comment]
@ellisonbg 1.4.1 came out the day this was merged. I'll submit a new PR with it soon.

**Comment by parente** [comment]
PR #60 has the bump.

**Comment by ellisonbg** [comment]
Many thanks!

On Mon, Jul 20, 2015 at 7:44 AM, Peter Parente <notifications@github.com>
wrote:

> PR #60 <https://github.com/jupyter/docker-demo-images/pull/60> has the
> bump.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jupyter/docker-demo-images/pull/58#issuecomment-122907072>
> .
>



-- 
Brian E. Granger
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

**Pull request merged**